Make dropdowns BS4 compliant

srvance opened the pull request:
Fixes #337. Continuing the conversation here.

I don't see a way to do this without a breaking change. That's not ideal since we've gone to beta, but I think it's necessary.

On rereading the docs, the `<a>` or `<button>` decision is about the dropdown button, not the item. For items, they only mention links or dividers. We have dividers covered, leaving us with two types of links, internal and external. I'm not worried separately about dropdown items that are plain text as that can be covered with a disabled item.

I'm going to create `bs-dropdown.menu.link-to` for the internal links.

External links need more thought as their tag structure between the two versions is significantly different.

simonihmig commented:
@srvance if I understand your intention correctly, you want to change this...

```hbs
{{#menu.item}}
  {{#link-to "index"}}Home{{/link-to}}
{{/menu.item}}
```

... to this...

```hbs
{{#menu.item}}
  {{#menu.link-to "index"}}Home{{/menu.link-to}}
{{/menu.item}}
```

Correct?

That does not need to be a breaking change, does it? For BS3 the first example would still work, but maybe not be the prefered way to do this, as it would prevent an easy upgrade path to BS4. And as BS4 support is still considered experimental, that would not be considered a breaking change, would it?

srvance commented:
I was thinking of changing

```handlebars
{{#menu.item}}
  {{#link-to "index"}}Home{{/link-to}}
{{/menu.item}}
```

to

```handlebars
{{#menu.link-to "index"}}Home{{/menu.link-to}}
```

but I like your suggestion better. It simplifies the change.

It's still breaking in that it requires a change of usage to be mutually compatible, but it's backwards compatible.

srvance commented:
@simonihmig Thank you for the insight. That simplified things tremendously. I was trying to wrap it all into `menu.item`.

@gijsbotje I can't explicitly request a review from you, but you should be able to contribute one. I would love it if you could point your app at this branch version of e-b and see if it addresses your concerns properly.

simonihmig commented on tests/integration/components/bs-tab-test.js:
```diff
@@ -142,12 +142,45 @@ test('tab navigation is groupable', function(assert) {
   ).textContent.trim(), 'Tab 3', 'dropdown menu item shows pane title');
   assert.equal(find(
-    'ul.nav.nav-tabs > li:nth-child(3) > ul.dropdown-menu > li:nth-child(2) > a'
+    'ul.nav.nav-tabs > li:nth-child(3) > .dropdown-menu > :nth-child(2) > a'
```
If you omit the last part of the selector (`> a`) here and on the statement before, this should still work, doesn't it? This is the only difference between the BS3 and BS4 tests, or did I miss something else? This way we could maybe unify the tests to avoid too much duplication?

srvance replied:
You're correct. The relaxed selector necessary to satisfy BS4 also works for BS3. The only disadvantage is that it's less specific about the DOM that's expected.

simonihmig commented on tests/integration/components/bs-dropdown/menu/link-to-test.js:
```diff
@@ -0,0 +1,32 @@
+testBS3('it has correct markup', function(assert) {
+  // Template block usage:
+  this.render(hbs`
+    {{#bs-dropdown/menu/link-to}}
```
Seems tests are failing because of the missing route parameter ("index" should work) for the link component!

simonihmig commented:
Oh btw, how do the changes here relate to #248, do they fix that issue as well?

gijsbotje commented:
my colleague @Rapid0o and were wondering why you wouldn't make a seperate v4 template
most of the bs4 elements have different structure from bs3
So wouldn't it be obvious to make seperate templates?

srvance commented:
@gijsbotje The differences have to do with component tags and classes, which don't require a template. I considered a template approach, which is what I was struggling with until @simonihmig's suggestion. The challenge there is coming up with a component definition that works for both BS3 and BS4 but still has all the flexibility of `{{link-to}}` and `<a>`. This would require accepting all attributes and passing them down appropriately, whereas this solution just requires slight changes to the component definitions and the new contextual definition.

srvance commented:
@simonihmig re #248, it might fix it. I'll have to check.

gijsbotje commented:
@srvance we looked at it again you're right :)
the changes to the tags and classnames is exactly what we where hoping for
@Rapid0o is pulling it to local as I type

Rapid0o commented:
@srvance These changes are exactly what we meant. It is according to the (alpha/beta) bs4 specs and addresses our concerns perfectly. You're right this is better than creating different templates for bs3 and bs4 :+1:

simonihmig commented:
@srvance Seems this is ready to go then! 🚀

What about this comment here: https://github.com/kaliber5/ember-bootstrap/pull/366#discussion_r122792275? This is not a major concern to me, but you might want to take a look? Otherwise changes are approved, so feel free to merge once you are happy with it! 😀

simonihmig commented:
@srvance Oh btw, when you merge, please squash the commits and create a commit message according to https://conventionalcommits.org/, (like `fix(Dropdown): some message`), so the changelog is populated automatically. I should write a `CONTRIBUTING.md` I guess...

srvance commented:
@simonihmig I'll take a look at the tab test comment this evening.
